lib/sdam/cursor.js

```diff
@@ -0,0 +1,747 @@
+ */
+function handleCallback(callback, err, result) {
+  try {
+    callback(err, result);
+  } catch (err) {
+    process.nextTick(function() {
+      throw err;
+    });
+  }
+}
+
+/**
+ * This is a cursor results callback
+ *
+ * @callback resultCallback
+ * @param {error} error An error object. Set to null if no error present
+ * @param {object} document
+ */
+
+/**
+ * An internal class that embodies a cursor on MongoDB, allowing for iteration over the
+ * results returned from a query.
+ *
+ * @property {number} cursorBatchSize The current cursorBatchSize for the cursor
+ * @property {number} cursorLimit The current cursorLimit for the cursor
+ * @property {number} cursorSkip The current cursorSkip for the cursor
+ */
+class Cursor {
+  /**
+   * Create a cursor
+   *
+   * @param {object} bson An instance of the BSON parser
+   * @param {string} ns The MongoDB fully qualified namespace (ex: db1.collection1)
+   * @param {{object}|Long} cmd The selector (can be a command or a cursorId)
+   * @param {object} [options=null] Optional settings.
+   * @param {object} [options.batchSize=1000] Batchsize for the operation
+   * @param {array} [options.documents=[]] Initial documents list for cursor
+   * @param {object} [options.transforms=null] Transform methods for the cursor results
+   * @param {function} [options.transforms.query] Transform the value returned from the initial query
+   * @param {function} [options.transforms.doc] Transform each document returned from Cursor.prototype.next
+   * @param {object} topology The server topology instance.
+   * @param {object} topologyOptions The server topology options.
+   */
+  constructor(bson, ns, cmd, options, topology, topologyOptions) {
+    options = options || {};
+
+    // Cursor pool
+    this.pool = null;
+    // Cursor server
+    this.server = null;
+
+    // Do we have a not connected handler
+    this.disconnectHandler = options.disconnectHandler;
+
+    // Set local values
+    this.bson = bson;
+    this.ns = ns;
+    this.cmd = cmd;
+    this.options = options;
+    this.topology = topology;
+
+    // All internal state
+    this.s = {
+      cursorId: null,
+      cmd: cmd,
+      documents: options.documents || [],
+      cursorIndex: 0,
+      dead: false,
+      killed: false,
+      init: false,
+      notified: false,
+      limit: options.limit || cmd.limit || 0,
+      skip: options.skip || cmd.skip || 0,
+      batchSize: options.batchSize || cmd.batchSize || 1000,
+      currentLimit: 0,
+      // Result field name if not a cursor (contains the array of results)
+      transforms: options.transforms
+    };
+
+    // Add promoteLong to cursor state
+    if (typeof topologyOptions.promoteLongs === 'boolean') {
+      this.s.promoteLongs = topologyOptions.promoteLongs;
+    } else if (typeof options.promoteLongs === 'boolean') {
+      this.s.promoteLongs = options.promoteLongs;
+    } else if (typeof options.session === 'object') {
```

Comment:
IIRC this is a bug, b/c it will not allow a session to attach if `promoteLongs` is present.

Reply:
nice catch! whoops.